neqo-client/src/main.rs

```diff
@@ -702,8 +711,22 @@ fn main() -> Res<()> {
             None
         }
     }) {
-        let addrs: Vec<_> = format!("{}:{}", host, port).to_socket_addrs()?.collect();
-        let remote_addr = *addrs.first().unwrap();
+        let addrs: Vec<_> = format!("{}:{}", host, port)
+            .to_socket_addrs()?
+            .filter(|addr| {
+                !matches!(
+                    (addr, args.ipv4_only, args.ipv6_only),
+                    (SocketAddr::V4(..), false, true) | (SocketAddr::V6(..), true, false)
+                )
+            })
+            .collect();
```

Comment:
You shouldn't need to collect() just to get the first() value.

```rust
        let remote_addr = format!("{}:{}", host, port)
            .to_socket_addrs()?
            .filter(|addr| {
                !matches!(
                    (addr, args.ipv4_only, args.ipv6_only),
                    (SocketAddr::V4(..), false, true) | (SocketAddr::V6(..), true, false)
                )
            })
            .first().expect(format!("No compatible address found for {}", host));
```
(That's a panic instead of eprintln!() and exit(), but you get the idea.  You can use `let remote_addr = if let Some(a) = ... .first() { *a } { eprintln!( ... ); exit(1); }` if you prefer.)

Reply:
Replaced with `find`. Kept the `match` separate as it's more readable.